lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -909,6 +995,41 @@ static NodeValue simplifyConcatNode(Function *F, ConcatNode *CN) {
     }
   }
 
+  /// If the inputs of concat are all reshape nodes, and the inputs of reshape
+  /// are able to be concatenated, then we can perform concat first and then
+  /// reshape. Concat(Reshape(x) * N) -> Reshape(Concat(x * N)).
+  {
+    std::vector<ReshapeNode *> reshapes;
+    for (auto &I : CN->getInputs()) {
+      if (auto *R = dyn_cast<ReshapeNode>(I)) {
+        // The input dims of Reshape nodes should have the same size.
+        if (reshapes.size() == 0 ||
+            reshapes.front()->getInput().dims().size() ==
+                R->getInput().dims().size()) {
+          reshapes.push_back(R);
+          continue;
+        }
+      }
+      // The inputs of Concat have to be all Reshape nodes. All the
+      // Reshape nodes need to also have the same shape.
+      return NodeValue(nullptr);
+    }
+
+    std::vector<NodeValue> newInputs;
+    for (auto &RN : reshapes) {
+      newInputs.push_back(RN->getInput());
+    }
+    // Check if the inputs are valid for a Concat node.
+    auto dim = getConcatDimensionForInputs(newInputs, CN);
+    if (dim == -1) {
+      return NodeValue(nullptr);
+    }
+    // Create the Concat node.
```

Comment:
make sure to remove all unnecessary comments.

Think of a comment as a way to clarify some complicated or non-obvious code.
Otherwise often times comment does not bring value and does not reflect the state correctly over time (gets stale).

Reply:
thx for pointing out